src/StatusManager.ts

```diff
@@ -0,0 +1,210 @@
+      this.updateStepInRoute(step)
+    }
+    return currentExecution
+  }
+
+  /**
+   * Updates the execution object of a Step.
+   * @param  {Step} step  The current step in execution
+   * @param  {Status} status  The status for the execution
+   * @param  {Receipt} receipt Optional. Information about received tokens
+   * @return {Step}       The step with the updated execution object
+   */
+  updateExecution(step: Step, status: Status, receipt?: Receipt): Step {
+    if (!step.execution) throw Error('Can update empty execution')
+    step.execution.status = status
+    if (receipt) {
+      step.execution.fromAmount = receipt.fromAmount
+      step.execution.toAmount = receipt.toAmount
+    }
+    this.updateStepInRoute(step)
+    return step
+  }
+
+  /**
+   * Create and push a new process into the execution. If a process for the given id already exists the existing process is set to PENDING instead.
+   * @param  {String} id  Identifier for the process. Used to identify already existing processes.
+   * @param  {Step} step The step that should contain the new process.
+   * @param  {ProcessMessage} message  A ProcessMessage for this Process. Will be used on newly created or already existing process.
+   * @param  {object} [params]   Additional parameters to append to the process.
+   * @return {Process}
+   */
+  findOrCreateProcess = (
+    id: string,
+    step: Step,
+    message: ProcessMessage,
+    params?: object
+  ): Process => {
+    if (!step.execution || !step.execution.process) {
+      throw new Error('Execution has not been initialized')
+    }
+
+    const process = step.execution.process.find((p) => p.id === id)
+
+    if (process) {
+      return process
+    }
+
+    const newProcess: Process = {
+      id: id,
+      startedAt: Date.now(),
+      message: message,
+      status: 'PENDING',
+    }
+    if (params) {
+      for (const [key, value] of Object.entries(params)) {
+        newProcess[key] = value
+      }
+    }
+
+    step.execution.process.push(newProcess)
+    this.updateStepInRoute(step)
+    return newProcess
+  }
+
+  /**
+   * Update a process object.
+   * @param  {Step} step The step where the process should be updated
+   * @param  {string} processId  The process id to update
+   * @param  {Status} status The status the process gets.
+   * @param  {object} [params]   Additional parameters to append to the process.
+   * @return {Process} The update process
+   */
+  updateProcess = (
+    step: Step,
+    processId: string,
+    status: Status,
+    params?: object
+  ): Process => {
+    const currentProcess = step?.execution?.process.find(
+      (p) => p.id === processId
+    )
+
+    if (!currentProcess) {
+      throw new Error('Cannot find process for given id')
+    }
+
+    switch (status) {
+      case 'CANCELLED':
+        currentProcess.doneAt = Date.now()
+        currentProcess.message =
+          'CANCELLED - Funds have been refunded on source chain.'
```

Comment:
I would leave it as it is for now since we only have this one usecase. 
As soon as we localize the frontend we will have to remove all those messages and replace them with codes anyways. Then the frontend can decide what to render for a `CANCELLED` status depending on the context. And it also has the information about the chain etc.